doc/contribute/coding_guidelines/index.rst

```diff
@@ -1297,6 +1297,101 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
 module, it is always possible to patch the macro implementation in each module
 to avoid collisions.
 
+.. _coding_guideline_libc_usage_restrictions_in_zephyr_kernel:
+
+Rule A.4: C Standard Library Usage Restrictions in Zephyr Kernel
+================================================================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+Zephyr kernel shall not use the C standard library functions and macros that
+are not available in the :ref:`minimal libc <c_library_minimal>`.
```

Comment:
I'd phrase this as using the standard C library as defined in ISO/IEC 9899:2018 plus a set of extensions (named here, identifying which standard they are defined by). That way we have both the list of functions *and* explicit specifications for them.

Reply:
In particular, I don't think this set of rules should refer to the minimal C library, it should define the Zephyr C library directly.

Reply:
At least for the Rule A.4, its purpose is "to ensure that the Zephyr kernel can build with the minimal libc," so I would prefer to explicitly refer to the set of functions available in the minimal libc.

As for the Rule A.5, which basically defines the "Zephyr C library API," I agree with you and will update the list to define the "specifications" for them.

Reply:
> At least for the Rule A.4, its purpose is "to ensure that the Zephyr kernel can build with the minimal libc," so I would prefer to explicitly refer to the set of functions available in the minimal libc.

That makes sense -- explicitly define the set of functions currently allowed for use by the Zephyr kernel. However, I think it should be worded the other way around -- define the set of functions allowed within the Zephyr kernel and then specify that the minimal C library must support all of those functions. When a new function is proposed for use within the kernel, this form of the specification would naturally drive the requirement that the PR include an implementation within the minimal C library.